amt/clients/clients.py

```diff
@@ -0,0 +1,92 @@
+import logging
+from abc import ABC, abstractmethod
+from datetime import datetime, timezone
+
+import httpx
+from amt.schema.github import RepositoryContent
+
+logger = logging.getLogger(__name__)
+
+
+class Client(ABC):
+    """
+    Abstract class which is used to set up HTTP clients that retrieve instruments from the
+    instrument register.
+    """
+
+    @abstractmethod
+    def __init__(self, max_retries: int = 3, timeout: int = 5) -> None:
+        transport = httpx.HTTPTransport(retries=max_retries)
+        self.client = httpx.Client(timeout=timeout, transport=transport)
+
+    @abstractmethod
+    def get_content(self, url: str) -> bytes:
+        """
+        This method should implement getting the content of an instrument from given URL.
+        """
+        ...
+
+    @abstractmethod
+    def list_content(self, url: str = "") -> RepositoryContent:
+        """
+        This method should implement getting list of instruments from given URL.
+        """
+        ...
+
+    def _get(self, url: str) -> httpx.Response:
+        """
+        Private function that performs a GET request to given URL.
+        """
+        response = self.client.get(url)
+        response.raise_for_status()
+        return response
+
+
+def get_client(repo_type: str) -> Client:
```

Comment:
Category "Nice to have".. but I just thought it might be an idea to use an Enum here for the types instead of string. For a "user" of the factory it is clear what options there are (without looking at the code)..

Reply:
Yes, 100% agree, but this would fail for Python 3.11 like we encountered last time ...